### ISSUE TITLE
Compare() in Shiny

### ISSUE THREAD

**Issue by isaacbain** [walkerke/mapgl#20]
Hello, 

I can't seem to get compare() working in a Shiny app. I've tried to make a minimal reproducible example here. It gives the error: `Warning in renderWidget(instance) :
  Ignoring explicitly provided widget ID "compare-container-db634"; Shiny doesn't use them`. 

```
# Load libraries
library(shiny)
library(mapgl)

# Define UI
ui <- fluidPage(
  titlePanel("Map Comparison"),
  mainPanel(
    mapboxglOutput("compare_map", height = "800px")
  )
)

# Define Server logic
server <- function(input, output, session) {
  
  output$compare_map <- renderMapboxgl({
    m1 <- mapboxgl(style = mapbox_style("light"))
    m2 <- mapboxgl(style = mapbox_style("dark"))
    
    # Use the compare function as documented
    compare(m1, m2)
  })
}

# Run the application
shinyApp(ui = ui, server = server)
```

**Comment by walkerke** [comment]
`compare()` doesn't work in Shiny yet.  Because of how it's structured it's technically a separate HTMLwidget that would need its own render function.  I'd like to get this working eventually!

**Comment by research-flow** [comment]
Hey, I would love to get this working, however I am not really familiarised with rendering HTMLwidgets yet. Are there any updates on this function? Thanks!

**Comment by timphan** [comment]
> `compare()` doesn't work in Shiny yet. Because of how it's structured it's technically a separate HTMLwidget that would need its own render function. I'd like to get this working eventually!

Hi Kyle, I have a similar question that might be related to this discussion. Recently you fixed the add_layers_control bug #80 (thanks for the quick turnaround!), and I noticed that whenever I tried doing a compare function with maps that had any additional widget (like layers control, draw control, etc) none of it would show up.

I'm assuming that based on your comment here - the compare function supersedes all existing HTML widgets? Is there a way to remedy this, I'd love to know what is possible.

Again, thank you for your time and great work on this package!

**Comment by walkerke** [comment]
Yeah this has proven to be a bit of a lift given how the compare widget works... technically it is a container for two maps, so this implementation would need to allow for interaction with the container itself _as well as_ the two maps inside the container.  

I may let Claude Code loose on this one and see what it can do!

**Comment by walkerke** [comment]
@timphan an update on this - it appears that the underlying library powering `compare()` for MapLibre does not support controls, but it does for Mapbox.  I've cleaned that up and fixed on the Mapbox side.  

The exceptions are controls custom to mapgl that aren't added via `map.addControl()` in JavaScript - this is the layers control and legend.  These _will_ work.

**Comment by walkerke** [comment]
Happy to share that this now works!  Use the new Shiny-specific rendering functions for `compare()`.

```r
# Load libraries
library(shiny)
library(mapgl)

# Define UI
ui <- fluidPage(
  titlePanel("Map Comparison"),
  mainPanel(
    mapboxglCompareOutput("compare_map")
  )
)

# Define Server logic
server <- function(input, output, session) {
  
  output$compare_map <- renderMapboxglCompare({
    m1 <- mapboxgl(style = mapbox_style("light"))
    m2 <- mapboxgl(style = mapbox_style("dark"))
    
    # Use the compare function as documented
    compare(m1, m2)
  })
}

# Run the application
shinyApp(ui = ui, server = server)
```

**Comment by walkerke** [comment]
> [@timphan](https://github.com/timphan) an update on this - it appears that the underlying library powering `compare()` for MapLibre does not support controls, but it does for Mapbox. I've cleaned that up and fixed on the Mapbox side.
> 
> The exceptions are controls custom to mapgl that aren't added via `map.addControl()` in JavaScript - this is the layers control and legend. These _will_ work.

Never mind -  controls now work in MapLibre compare!
